File: backend/app/schemas/release_gate.py
# CRANE — CRA Norm Engine
# Copyright (C) 2026 Ali Mohammad Hosseini
# SPDX-License-Identifier: AGPL-3.0-or-later
#
# This file is part of CRANE, free software under the GNU Affero General Public
# License v3.0 or later. See <https://www.gnu.org/licenses/>.

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ArtifactReviewDecision, ReleaseGateItemCode, ReleaseGateWorkflowStatus
from app.schemas.artifact import ArtifactRevisionRead
from app.schemas.product_release import ProductReleaseRead
from app.schemas.user import UserSummaryRead


class ReleaseGateReviewRead(BaseModel):
    """One historical reviewer decision + note on an evidence link."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    decision: ArtifactReviewDecision
    rationale: str | None
    reviewed_by_user: UserSummaryRead | None = None
    reviewed_at: datetime


class ReleaseGateEvidenceLinkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    decision: ArtifactReviewDecision
    rationale: str | None
    linked_by_user_id: UUID
    linked_by_user: UserSummaryRead | None = None
    reviewed_by_user_id: UUID | None
    reviewed_by_user: UserSummaryRead | None = None
    reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    reviews: list[ReleaseGateReviewRead] = []
    artifact_revision: ArtifactRevisionRead


class ReleaseGateItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: ReleaseGateItemCode | None = None
    title: str
    description: str | None
    is_required: bool
    sort_order: int
    status: ArtifactReviewDecision
    assigned_to_user_id: UUID | None = None
    due_date: datetime | None = None
    evidence_links: list[ReleaseGateEvidenceLinkRead] = []


class ReleaseGateRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    product_release_id: UUID
    status: ReleaseGateWorkflowStatus
    submitted_at: datetime | None
    submitted_by_user_id: UUID | None
    submitted_by_user: UserSummaryRead | None = None
    approved_at: datetime | None
    approved_by_user_id: UUID | None
    approved_by_user: UserSummaryRead | None = None
    bundle_sha256: str | None = None
    bundle_generated_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    items: list[ReleaseGateItemRead] = []
    required_items_count: int = 0
    accepted_items_count: int = 0
    pending_items_count: int = 0


class ReleaseGateDetailRead(BaseModel):
    release: ProductReleaseRead
    gate: ReleaseGateRead


class ReleaseGateReviewRequest(BaseModel):
    decision: ArtifactReviewDecision
    rationale: str | None = Field(default=None, max_length=4000)


class GateItemCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)

File: frontend/src/types/release-gate.ts
import type { ArtifactRevisionRead, UserSummaryRead } from "@/types/artifact";
import type { ConformityRoute, ProductClassification } from "@/types/product";

export type ReleaseGateStatus = "draft" | "in_review" | "approved" | "blocked";
export type GateDecision = "pending_review" | "accepted" | "rejected" | "needs_update" | "waived";
export type ReleaseGateItemCode =
  | "technical_documentation"
  | "risk_assessment"
  | "sbom"
  | "test_report"
  | "declaration_of_conformity"
  | "annex_mapping";

export interface ProductReleaseRead {
  id: string;
  product_id: string;
  product_name: string | null;
  system_version: number;
  system_version_label: string;
  user_version: string | null;
  display_version: string;
  release_status:
    | "draft"
    | "in_review"
    | "blocked"
    | "approved"
    | "placed_on_market"
    | "released"
    | "withdrawn"
    | "recalled"
    | "end_of_support";
  planned_release_date: string | null;
  actual_release_date: string | null;
  placed_on_market_date: string | null;
  classification_snapshot: ProductClassification;
  conformity_route_snapshot: ConformityRoute;
  release_notes: string | null;
  parent_release_id: string | null;
  substantiality_analysis_id: string | null;
  is_consolidated_support_version: boolean;
  // CRA Art. 13(8) traceability: ID of the substantial change that required this re-release.
  caused_by_change_id: string | null;
  /** CRA Art. 13(2): true when this release has known exploitable vulnerabilities blocking gate approval. */
  has_known_exploitable_vulnerabilities: boolean;
  kev_notes: string | null;
  eu_doc_date: string | null;
  eu_doc_number: string | null;
  eu_doc_notified_body: string | null;
  /** Gap 2 — hardware version for embedded products. */
  hardware_version: string | null;
  /** Gap 2 — software/firmware version for embedded products. */
  software_version: string | null;
  /** Gap 1 — remote processing elements linked to this release. */
  release_remote_processing_elements: import("@/types/product").RemoteProcessingElementSummaryRead[];
  created_at: string;
  updated_at: string;
}

export interface ReleaseGateEvidenceLinkRead {
  id: string;
  decision: GateDecision;
  rationale: string | null;
  linked_by_user_id: string;
  linked_by_user: UserSummaryRead | null;
  reviewed_by_user_id: string | null;
  reviewed_by_user: UserSummaryRead | null;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
  artifact_revision: ArtifactRevisionRead;
}

export interface ReleaseGateItemRead {
  id: string;
  code: ReleaseGateItemCode | null;
  title: string;
  description: string | null;
  is_required: boolean;
  sort_order: number;
  status: GateDecision;
  evidence_links: ReleaseGateEvidenceLinkRead[];
  prerequisites?: ReleaseGateItemSummary[];
  unmet_prerequisites?: ReleaseGateItemSummary[];
}

export interface ReleaseGateItemSummary {
  id: string;
  code: ReleaseGateItemCode | null;
  title: string;
  status: GateDecision;
}

export interface ReleaseGateRead {
  id: string;
  product_release_id: string;
  status: ReleaseGateStatus;
  submitted_at: string | null;
  submitted_by_user_id: string | null;
  submitted_by_user: UserSummaryRead | null;
  approved_at: string | null;
  approved_by_user_id: string | null;
  approved_by_user: UserSummaryRead | null;
  bundle_sha256: string | null;
  bundle_generated_at: string | null;
  snapshot_json: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  items: ReleaseGateItemRead[];
  required_items_count: number;
  accepted_items_count: number;
  pending_items_count: number;
}

export interface ReleaseGateDetailRead {
  release: ProductReleaseRead;
  gate: ReleaseGateRead;
}
